### PR TITLE
Update with all *real* models

### PR DESCRIPTION
**Pull request by jdiaz4302**
Just prior to this PR I merged a dummy PR into main which duplicated the `02_train/src/A` notebook into an A, B, C, D, E, and F version. This PR is about adding the actual models into those dummy duplicates.

# Overview

`A`, `C`, and `E` are LSTM notebooks. Later letters correspond to larger model sizes: `A` = average, `C` = 3 million parameters, and `E` = 6 layers, 3 million parameters each, 18 million total. `B`, `D`, and `F` are transformer notebooks; their ordered letters follow the same size scheme.

* `A` and `B` were chosen to match the approximate normal size of an LSTM in hydrology/IIDD work.
* `F` is the original size of the full encoder (6 layers) in the original transformer paper. `E` is an LSTM with equivalent layers and total number of parameters.
  * There is no autoregression in this lake ice prediction (i.e., I'm not using past lake ice values to predict future lake ice), so an encoder-only approach seemed most applicable as a transformer model. This essentially replaces recurrence with self-attention on the exogenous variables. 
* `C` and `D` are 1-layer equivalents of `E` and `F`.

# Caveat

Not much actually changes between these notebooks. Realistically, these 6 notebooks could be 1 notebook, but it was easier during development to have 6 notebooks rather than 1 really flexible notebook. If any substantive changes need to made, I will likely consolidate them into 1 notebook, but I would rather make as little polishing/design changes as necessary at this point.

# Tips for reviewing

Much of what changes in each notebook is:

* Input values - batch size (to maximize GPU RAM), number of parameters/hidden layer size, number of layers, and file paths (e.g., `massive_lstm`)
* Model code - this is largely identical within model types. The "massive" notebooks are altered to allow a specification of number of layers
* Model definition (e.g., specifying number of layers)

# Requested review

Identification of bugs or erroneous implementation. Everything seems to be working as expected for many iterations at this point, but with transformers being new to us, it may be worth taking a closer look at their implementation and usage.

### CODE REVIEW

File: 02_train/src/B_FitBasicTransformer.ipynb [hunk 1]
```diff
@@ -244,41 +247,83 @@
...
+    "        pe = torch.moveaxis(pe, [0, 1, 2], [1, 0, 2]) # BATCH FIRST\n",
+    "        self.register_buffer('pe', pe)\n",
+    "\n",
+    "    def forward(self, x: Tensor) -> Tensor:\n",
+    "        \"\"\"\n",
+    "        Args:\n",
+    "            x: Tensor, shape [seq_len, batch_size, embedding_dim]\n",
+    "        \"\"\"\n",
+    "        x = x + self.pe[:x.size(0)]\n",
+    "        return self.dropout(x)"
```
Comment (jdiaz4302) [inline 02_train/src/B_FitBasicTransformer.ipynb:280]: This is a standard sinusoidal positional encoding taken from https://pytorch.org/tutorials/beginner/transformer_tutorial.html

File: 02_train/src/B_FitBasicTransformer.ipynb [hunk 2]
```diff
@@ -244,41 +247,83 @@
...
+    "        self.mask = self.mask.transpose(0, 1)\n",
+    "        self.mask = self.mask.cuda()\n",
     "        \n",
-    "        lstm_out, (h, c) = self.lstm(x)\n",
+    "        self.encoder = nn.TransformerEncoderLayer(d_model = model_dim,\n",
+    "                                                  nhead = nheads,\n",
+    "                                                  batch_first = True,\n",
+    "                                                  dim_feedforward = ff_dim)\n",
+    "        self.dense = nn.Linear(model_dim, 1)\n",
+    "        self.dense_activation = nn.Sigmoid()\n",
```
Comment (jdiaz4302) [inline 02_train/src/B_FitBasicTransformer.ipynb:310]: Model structure:

- Embedding (dense layer) to project the input data to the transformer model size (functions similarly to a word embedding generator)
- A mask is used to prevent looking into the future
- A transformer encoder with self-attention is used to transform the embedding
- A dense layer acts on the encoder's output

During the forward pass, the positional encoding is added prior to the encoder.

File: 02_train/src/B_FitBasicTransformer.md [hunk 3]
```diff
@@ -139,27 +142,63 @@ for i in range(train_x.shape[2]):
...
+        """Assumes x is of shape (batch, sequence, feature)"""
         
-        out = self.activation(self.dense(drop_out))
+        embed = self.embedding(x)
+        position = self.pos_encoding(embed)
+        embed = embed + position
+        encoded = self.encoder(embed, src_mask = self.mask)
+        out = self.dense_activation(self.dense(encoded))
         
         return out
```
Comment (jdiaz4302) [inline 02_train/src/B_FitBasicTransformer.md:203]: Model structure:

- Embedding (dense layer) to project the input data to the transformer model size (functions similarly to a word embedding generator)
- A mask is used to prevent looking into the future
- A transformer encoder with self-attention is used to transform the embedding
- A dense layer acts on the encoder's output

During the forward pass, the positional encoding is added prior to the encoder.

File: 02_train/src/B_FitBasicTransformer.md [hunk 4]
```diff
@@ -139,27 +142,63 @@ for i in range(train_x.shape[2]):
...
+        pe = torch.moveaxis(pe, [0, 1, 2], [1, 0, 2]) # BATCH FIRST
+        self.register_buffer('pe', pe)
+
+    def forward(self, x: Tensor) -> Tensor:
+        """
+        Args:
+            x: Tensor, shape [seq_len, batch_size, embedding_dim]
+        """
+        x = x + self.pe[:x.size(0)]
+        return self.dropout(x)
```
Comment (jdiaz4302) [inline 02_train/src/B_FitBasicTransformer.md:168]: This is a standard sinusoidal positional encoding taken from https://pytorch.org/tutorials/beginner/transformer_tutorial.html

File: 02_train/src/C_FitBasicLSTM_Large.ipynb [hunk 5]
```diff
@@ -77,15 +77,18 @@
...
-    "coarse_epoch_printing = 1000\n",
+    "coarse_epoch_printing = 50\n",
     "\n",
     "# model hyperparams\n",
     "random_seed = 4 # change for different 'random' initializations\n",
-    "model_dim = 16\n",
+    "params = 3159041 # matching the full size, encoder only, transformer\n",
+    "model_dim = int(np.round((1/8)*(np.sqrt(16*params + 2385) - 49))) # assumes 11 variables \n",
+    "# ^ solves for y where y = 4x^2 + 49x + 1\n",
+    "# which was originally y = 11*4*x + 4*x*x + 4*x + 1*x + 1\n",
```
Comment (jdiaz4302) [inline 02_train/src/C_FitBasicLSTM_Large.ipynb:87]: Glimpse at the math used to match LSTM and transformer model size

File: 02_train/src/C_FitBasicLSTM_Large.md [hunk 6]
```diff
@@ -49,15 +49,18 @@ valid_data_fpath = extended_dir + valid_data_fpath
...
-coarse_epoch_printing = 1000
+coarse_epoch_printing = 50
 
 # model hyperparams
 random_seed = 4 # change for different 'random' initializations
-model_dim = 16
+params = 3159041 # matching the full size, encoder only, transformer
+model_dim = int(np.round((1/8)*(np.sqrt(16*params + 2385) - 49))) # assumes 11 variables 
+# ^ solves for y where y = 4x^2 + 49x + 1
+# which was originally y = 11*4*x + 4*x*x + 4*x + 1*x + 1
```
Comment (jdiaz4302) [inline 02_train/src/C_FitBasicLSTM_Large.md:59]: Glimpse at the math used to match LSTM and transformer model size

Comment (jdiaz4302) [reply 02_train/src/C_FitBasicLSTM_Large.md:59]: Subsequent prints verify that the number of parameters is highly comparable (can't literally match due to different structures, but its ~99% matched)

File: 02_train/src/E_FitBasicLSTM_Massive.md [hunk 7]
```diff
@@ -49,15 +49,19 @@ valid_data_fpath = extended_dir + valid_data_fpath
...
 
 # model hyperparams
 random_seed = 4 # change for different 'random' initializations
-model_dim = 16
+params = 18920961 # matching the full size, encoder only, transformer
+model_dim = int(np.round((1/88)*(np.sqrt(176*params + 4585) - 69))) # assumes 11 variables 
+# ^ solves for y where y = 4x^2 + 49x + 1
+# which was originally y = 11*4*x + 4*x*x + 4*x + 1*x + 1
 dropout_val = 0.1 # matching encoder default value
+nlayers = 6
```
Comment (AndyMcAliley) [inline 02_train/src/E_FitBasicLSTM_Massive.md:61]: Woo, this one is tougher. So, six layers. The first layer has 11 features, so it's like before: 4x(11+x) weights and 4x biases. The other five layers have x features, so they have 4x(x+x) weights and 4x biases. Then there's the final densely connected layer with x*1 weights and 1 bias. In total,

$y = 4x(11+x)+4x+5(4x(x+x)+4x)+1x+1$
$y = 44x+4x^2+4x+5(8x^2+4x)+1x+1$
$y = 44x^2+69x+1$

Solving for x,

$x = \frac{1}{2(44)}(-69 \pm \sqrt{69^2-4(44)(1-y))}$

The positive root is

$x=\frac{1}{88}(\sqrt{4761-176+176y}-69)$
$x=\frac{1}{88}(\sqrt{4585+176y}-69)$

Looks good!

File: 02_train/src/F_FitBasicTransformer_Massive.md [hunk 8]
```diff
@@ -139,27 +143,66 @@ for i in range(train_x.shape[2]):
...
+        self.dense_activation = nn.Sigmoid()
         
-        drop_out = self.dropout(lstm_out)
+    def forward(self, x):
+        """Assumes x is of shape (batch, sequence, feature)"""
         
-        out = self.activation(self.dense(drop_out))
+        embed = self.embedding(x)
+        position = self.pos_encoding(embed)
+        embed = embed + position
```
Comment (AndyMcAliley) [inline 02_train/src/F_FitBasicTransformer_Massive.md:202]: ...And again, tagging just in case.

**Review by AndyMcAliley — CHANGES_REQUESTED** [review]
Looks good! See my comment in the forward method of the BasicTransformer class for one implementation detail to double check before merging.

File: 02_train/src/B_FitBasicTransformer.md (hunk 4 above)
Comment (jdiaz4302) [reply 02_train/src/B_FitBasicTransformer.md:168]: `nn.TransformerEncoderLayer` uses a default dropout value of `0.1` which I did not override, and I used this value for some small regularization with the LSTMs as well.

https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoderLayer.html

Comment (jdiaz4302) [reply 02_train/src/B_FitBasicTransformer.md:168]: I saw no harm leaving the dropout there (in the positional encoding) since it seems to fit a broader pattern of using lots of low-magnitude dropout in most transformer layers

Comment (AndyMcAliley) [reply 02_train/src/B_FitBasicTransformer.md:168]: Sounds good!

**Review by AndyMcAliley — DISMISSED** [review]
Looks good!

**Review by AndyMcAliley — APPROVED** [review]
